### PR TITLE
Improvement: Option for Guests to use more of wide path

### PR DESCRIPTION
**Pull request by WantDiscussion**
Improvement: Option for Guests to use more of wide path

-Instead of only one wide flag, each path can have multiple levels of flags

-The first level runs the basic algorithm 8 times (or "groups") in
multiple directions for peeps to use paths symmetrically.

-Levels 2 and above each has 8 groups coresponding to the previous flags
to fill in the remaining spaces (Still incomplete)

-When peeps are path finding, the flag they consider when checking if a
path is wide is based on their ID. The original wide flag still keeps
track of the default update order. Staff peeps use the default wide
Flag.

-Each group and level of flags can be viewed with the debugger.

-Flags are stored in a array of vectors separate from the
tile structures

-In place of saving all flags, when loading a map the game will instead
calculate all wide flags when configured to use 1 or more levels. This means
the game state is not perfectly preserved upon saving when this feature
is on.

Future Improvements:
-Improve algorithm for widepath levels 2 and up. Some gaps are not yet covered.

### CODE REVIEW

File: src/openrct2/world/TileElement.h [hunk 1]
```diff
@@ -272,7 +277,7 @@ struct PathElement : TileElementBase
 
     bool IsLevelCrossing(const CoordsXY& coords) const;
 };
-assert_struct_size(PathElement, 16);
+assert_struct_size(PathElement, 17);
```
Comment (janisozaur) [inline src/openrct2/world/TileElement.h]: I doubt this will pass. These asserts are here to maintain SV6 compatibility. You can see a number of recent contributions to see how to deal with that: read old structs from files, use them to set new ones.

Comment (duncanspumpkin) [reply src/openrct2/world/TileElement.h]: Its not to maintain SV6 compatibility as that was 8 bytes. These asserts are because we don't want this struct to exceed 16 bytes. So its still the same outcome in that we can't make this change.

File: src/openrct2/peep/Staff.cpp [hunk 2]
```diff
@@ -293,7 +293,6 @@ bool staff_can_ignore_wide_flag(Peep* staff, int32_t x, int32_t y, uint8_t z, Ti
                 pathfound = true;
                 pathcount++;
             }
-
```
Comment (tupaschoal) [inline src/openrct2/peep/Staff.cpp]: This is unrelated, so please revert.

File: src/openrct2/world/TileElement.h [hunk 3]
```diff
@@ -199,15 +199,17 @@ struct PathElement : TileElementBase
...
     uint8_t Flags2;         // 9
     uint8_t SlopeDirection; // 10
+
     union
     {
         uint8_t AdditionStatus; // 11, only used for litter bins
         ride_id_t rideIndex;    // 11
     };
     ::StationIndex StationIndex; // 13
+    uint8_t isWideFlags;         // 14
```
Comment (tupaschoal) [inline src/openrct2/world/TileElement.h]:
```suggestion
    uint8_t WideFlags;         // 14
```
I think the "is" should be dropped. "is..." variables usually imply boolean value.

File: src/openrct2/paint/tile_element/Paint.Path.cpp [hunk 4]
```diff
@@ -850,7 +850,7 @@ void path_paint(paint_session* session, uint16_t height, const TileElement* tile
     }
 
     // Draw wide flags as ghosts, leaving only the "walkable" paths to be drawn normally
-    if (gPaintWidePathsAsGhost && tile_element->AsPath()->IsWide())
+    if (gPaintWidePathsAsGhost && tile_element->AsPath()->IsWideForGroup(gDateMonthTicks / 8192))
```
Comment (tupaschoal) [inline src/openrct2/paint/tile_element/Paint.Path.cpp]: What's with this division?

File: src/openrct2/world/Footpath.cpp [hunk 5]
```diff
@@ -1630,7 +1647,11 @@ static void footpath_clear_wide(const CoordsXY& footpathPos)
     {
         if (tileElement->GetType() != TILE_ELEMENT_TYPE_PATH)
             continue;
-        tileElement->AsPath()->SetWide(false);
+        tileElement->AsPath()->SetWideForGroup(wideGroup, false);
+
+        if (wideGroup == 0 && !(tileElement->AsPath()->IsWideForGroup(0)))
```
Comment (tupaschoal) [inline src/openrct2/world/Footpath.cpp]:
```suggestion
        if (wideGroup == 0 && !tileElement->AsPath()->IsWideForGroup(wideGroup))
```

**Comment by AaronVanGeffen** [comment]
Hey @WantDiscussion,

Thank you for your working on improving the pathfinding. We know the current pathing algorithm is suboptimal. Fortunately, various contributors have worked hard to improve these algorithms before, from ironing out obvious bugs to documenting existing behaviour. This has made them better than vanilla RCT2. However, there is certainly room for improvement.

At present, we are only considering improvements to the pathfinding algorithm that are bugfixes. By this I mean changes that are relatively self-contained. I speak for the team when I say that, unfortunately, that means we have decided not accept this PR at this time.

I am very sorry to break this news to you, because you have obviously been working very hard on this PR.

The next step on our roadmap for the peep pathfinding is to make the entire system pluggable. In this way, players could choose to experience the vanilla pathfinding algorithm, or an entirely different algorithm. Such an algorithm could trade more optimal pathing for more intensive CPU usage, for example.

This is quite an undertaking, however, and work on this is on hold as other changes are prioritised (the new save format and plug-in engine, for example). However, until this is done, we will reject PRs like #3767, #6440, and this one.

If you have smaller, self-contained changes to improve the pathing algorithm, I would like to encourage you to extract them into a new PR. However, please keep the above in mind. In general, it is good practice to ask around on Discord or Gitter before undertaking big changes.

Finally, I would like to point out that this decision is not meant, in any way, to reflect your coding abilities. Indeed, it is obvious that you worked very hard and enthusiastically on tackling this. I hope this decision will not discourage you from future contributions.

**Comment by WantDiscussion** [comment]
Thank you for your response.
